File: src/main/kotlin/org/example/schwimmen/eingabe/Staffeln.kt
package org.example.schwimmen.eingabe

import org.example.schwimmen.model.Staffel

fun parseStaffeln(data: String): List<Staffel> = parseStaffeln(data.lines().map { it.split("\t") })

fun parseStaffeln(rows: List<List<String>>): List<Staffel> {
    val result = mutableListOf<Staffel>()
    var staffelBuilder: StaffelBuilder? = null
    for (row in rows) {
        if (isHeader(row)) {
            staffelBuilder?.let { result.add(it.build()) }
            staffelBuilder = StaffelBuilder(row[1], mutableListOf())
        } else if (row[0].isNotBlank()) {
            (staffelBuilder ?: error("Konfigurationsformat falsch. Header 'Staffel' nicht gefunden.")).startsDisziplinen.add(
                Pair(row[0].toInt(), row[1]),
            )
        }
    }

    staffelBuilder?.let { result.add(it.build()) }

    return result
}

private fun isHeader(row: List<String>): Boolean = row[0].trim() == "Staffel"

data class StaffelBuilder(
    val name: String,
    val startsDisziplinen: MutableList<Pair<Int, String>>,
) {
    fun build(): Staffel =
        Staffel(
            name,
            startsDisziplinen.flatMap { (starts, disziplin) -> List(starts) { disziplin } },
            team = name.contains("team", ignoreCase = true),
        )
}

File: web/src/lib/schwimmen/eingabe/staffeln.ts
export interface Staffel {
  name: string;
  disziplinen: string[];
  team: boolean;
}

export function parseStaffeln(data: string): Staffel[] {
  return parseStaffelnFromGrid(data.split("\n").map((row) => row.split("\t")));
}

export function parseStaffelnFromGrid(rows: string[][]): Staffel[] {
  const result: Staffel[] = [];
  let staffelBuilder: StaffelBuilder | undefined = undefined;
  for (const row of rows) {
    if (isHeader(row)) {
      if (staffelBuilder) {
        result.push(buildStaffel(staffelBuilder));
      }
      staffelBuilder = { name: row[0], startsDisziplinen: [] };
    }

    if (row[1].trim().length != 0) {
      if (!staffelBuilder) {
        throw Error("Konfigurationsformat falsch. Header 'Staffel' nicht gefunden.");
      }
      staffelBuilder.startsDisziplinen.push({ starts: Number(row[1]), disziplin: row[2] });
    }
  }

  if (staffelBuilder) {
    result.push(buildStaffel(staffelBuilder));
  }

  return result;
}

function isHeader(row: string[]): boolean {
  return row[0].trim().length > 0;
}

interface StaffelBuilder {
  name: string;
  startsDisziplinen: { starts: number; disziplin: string }[];
}
function buildStaffel(staffelBuilder: StaffelBuilder): Staffel {
  return {
    name: staffelBuilder.name,
    disziplinen: staffelBuilder.startsDisziplinen.flatMap(({ starts, disziplin }) => Array(starts).fill(disziplin)),
    team: staffelBuilder.name.toLowerCase().includes("team"),
  };
}
